From 1947bb0e260a0ee61f64514187d5494fff057d06 Mon Sep 17 00:00:00 2001
From: Fancy Fang <chen.fang@nxp.com>
Date: Mon, 7 Aug 2017 13:53:50 +0800
Subject: [PATCH 2335/5242] MLK-16136-14 video: fbdev: dcss: refine ctxld
 interrupts unmask logic

commit  2342ec24d89774e2dd3cdf5d0ad9b68a151f256e from
https://source.codeaurora.org/external/imx/linux-imx.git

The ctxld interrupts are better to be unmasked during
interrupts initialization in probe stage with only once.

Signed-off-by: Fancy Fang <chen.fang@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/video/fbdev/mxc/imx_dcss.c |   15 ++++++++++-----
 1 file changed, 10 insertions(+), 5 deletions(-)

diff --git a/drivers/video/fbdev/mxc/imx_dcss.c b/drivers/video/fbdev/mxc/imx_dcss.c
index f81889e..0e794b0 100644
--- a/drivers/video/fbdev/mxc/imx_dcss.c
+++ b/drivers/video/fbdev/mxc/imx_dcss.c
@@ -2418,11 +2418,6 @@ static void dcss_ctxld_config(struct work_struct *work)
 		       info->base + chans->ctxld_addr + CTXLD_DB_COUNT);
 	}
 
-	ctxld_irq_unmask(DB_COMP_EN, info);
-	ctxld_irq_unmask(SB_HP_COMP_EN, info);
-	ctxld_irq_unmask(SB_LP_COMP_EN, info);
-	ctxld_irq_unmask(AHB_ERR_EN, info);
-
 	ctxld_ctrl = readl(info->base + chans->ctxld_addr + CTXLD_CTRL_STATUS);
 	ctxld_ctrl |= (1 << 0);
 
@@ -2904,6 +2899,7 @@ static irqreturn_t dcss_irq_handler(int irq, void *dev_id)
 static int dcss_interrupts_init(struct dcss_info *info)
 {
 	int i, ret = 0;
+	struct irq_desc *desc;
 	struct platform_device *pdev = info->pdev;
 
 	for (i = 0; i < DCSS_IRQS_NUM; i++) {
@@ -2911,6 +2907,15 @@ static int dcss_interrupts_init(struct dcss_info *info)
 		if (info->irqs[i] < 0)
 			break;
 
+		desc = irq_to_desc(info->irqs[i]);
+		switch (desc->irq_data.hwirq) {
+		case 6:         /* CTX_LD */
+			ctxld_irq_unmask(SB_HP_COMP_EN, info);
+			break;
+		default:	/* TODO: add support later */
+			continue;
+		}
+
 		ret = devm_request_irq(&pdev->dev, info->irqs[i],
 				dcss_irq_handler, 0,
 				dev_name(&pdev->dev), info);
-- 
1.7.9.5

